autopilot/agent_test.go

```diff
@@ -1091,3 +1091,219 @@ func TestAgentQuitWhenPendingConns(t *testing.T) {
+
+	select {
+	case req := <-testCtx.heuristic.nodeScoresArgs:
+		// All nodes in the graph should be potential channel
+		// candidates.
+		if len(req.nodes) != len(nodeScores) {
+			t.Fatalf("expected %v nodes, instead had %v",
+				len(nodeScores), len(req.nodes))
+		}
+
+		// One channel is already open.
+		if len(req.chans) != int(existingChans) {
+			t.Fatalf("expected 1 existing channel, got %v",
+				len(req.chans))
+		}
+		if req.amt != chanSize {
+			t.Fatalf("expected channel size of %v, got %v",
+				chanSize, req.amt)
+		}
+
+	case <-time.After(time.Second * 3):
+		t.Fatalf("select wasn't queried in time")
+	}
+
+	// Respond with the given scores.
+	select {
+	case testCtx.heuristic.nodeScoresResps <- nodeScores:
+	case <-time.After(time.Second * 3):
+		t.Fatalf("NodeScores wasn't queried in time")
+	}
+}
+
+// checkChannelOpens asserts that the channel controller attempts open the
+// number of channels we expect, and with the exact total allocation.
+func checkChannelOpens(t *testing.T, testCtx *testContext,
+	allocation btcutil.Amount, numChans int) []NodeID {
+
+	var nodes []NodeID
+
+	// The agent should attempt to open channels, totaling what we expect.
+	var totalAllocation btcutil.Amount
+	chanController := testCtx.chanController.(*mockChanController)
+	for i := 0; i < numChans; i++ {
+		select {
+		case openChan := <-chanController.openChanSignals:
+			totalAllocation += openChan.amt
+
+			testCtx.Lock()
+			testCtx.walletBalance -= openChan.amt
+			testCtx.Unlock()
+
+			nodes = append(nodes, NewNodeID(openChan.target))
+
+		case <-time.After(time.Second * 3):
+			t.Fatalf("channel not opened in time")
+		}
+	}
+
+	if totalAllocation != allocation {
+		t.Fatalf("expected agent to open channels totalling %v, "+
+			"instead was %v", allocation, totalAllocation)
+	}
+
+	// Finally, make sure the agent won't try opening more channels.
+	select {
+	case <-chanController.openChanSignals:
+		t.Fatalf("agent unexpectedly opened channel")
+
+	case <-time.After(50 * time.Millisecond):
+	}
+
+	return nodes
+}
+
+// TestAgentChannelSizeAllocation tests that the autopilot agent opens channel
+// of size that stays within the channel budget and size restrictions.
+func TestAgentChannelSizeAllocation(t *testing.T) {
+	t.Parallel()
+
+	// Total number of nodes in our mock graph.
+	const numNodes = 10
+
+	testCtx, cleanup := setup(t, nil)
+	defer cleanup()
+
+	nodeScores := make(map[NodeID]*NodeScore)
+	for i := 0; i < numNodes; i++ {
+		nodeKey, err := testCtx.graph.addRandNode()
+		if err != nil {
+			t.Fatalf("unable to generate key: %v", err)
+		}
+		nodeID := NewNodeID(nodeKey)
+		nodeScores[nodeID] = &NodeScore{
+			NodeID: nodeID,
+			Score:  0.5,
+		}
+	}
+
+	// The agent should now query the heuristic in order to determine its
+	// next action as it local state has now been modified.
+	select {
+	case arg := <-testCtx.constraints.moreChanArgs:
+		if len(arg.chans) != 0 {
+			t.Fatalf("expected agent to have 1 channel already "+
```

Comment:
Should be `expected agent to not have any open channels`?

Reply:
Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/30bf23d9256a5d2dec7fb980b23bf9434c3196e0